juju/juju
# provider/gce: Don't base the name of firewall rules on hash of CIDRs

## Description of change

Found when testing #7234 - the updated addresses would be logged for updating the firewall, but the update wouldn't happen. The error on updating was being swallowed somewhere so it wasn't clear what was happening from the logs. (I'll fix that in a separate PR when I track it down.)

We were generating the name of the firewall using a hash of the new rules, and trying to update the firewall to have the new name. GCE will prevent a mismatch between the URL and the name with this
error:

```
opening port(s) [3306/tcp from 35.187.146.143/32]: googleapi: Error 400: Invalid value for field 'resource.name': 'juju-893de4-0-d4bc50'. The name of the resource in request body must be the same as the resource name in the URL, invalid
```
Since we can't update the name of the firewall rules it seems like the options are:

1. Create a new one (with the name still based on the hashed CIDRs) and then remove the old one.
2. Change the naming to not be based on hashing the CIDRs - use a random unique string instead.

Option 1 can't be done atomically, and cleaning up the mess if the create happened but the delete didn't seems like it would be nasty. So this change implements option 2.

Since the names will be unpredictable we need to hang on to them when we get the existing rules from GCE so we can send them back with updates/deletes. The old version of the code wasn't doing that - names were thrown away when the compute.Firewall is converted into []network.IngressRule because it could regenerate the name from the CIDRs.


## QA steps

* Bootstrap on GCE, create two models with a cross-model relation between mysql and wordpress. 
* Expose wordpress
* Check that you can successfully see the wordpress setup page at the wordpress public address.
* In the GCE console, stop the wordpress machine and then start it again - note that its public address has changed.
* Wait for the Juju instance poller to notice the updated public address.
* Check that the wordpress setup page is available at the new address.
